x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/aggs/kstest/BucketCountKSTestAggregationBuilder.java

```diff
@@ -0,0 +1,195 @@
+
+    public static final ParseField NAME = new ParseField("bucket_count_ks_test");
+    @SuppressWarnings("unchecked")
+    public static final ConstructingObjectParser<BucketCountKSTestAggregationBuilder, String> PARSER = new ConstructingObjectParser<>(
+        NAME.getPreferredName(),
+        false,
+        (args, context) -> new BucketCountKSTestAggregationBuilder(
+            context,
+            (String) args[0],
+            (List<Double>) args[1],
+            (List<String>) args[2],
+            (BucketHelpers.GapPolicy) args[3],
+            (SamplingMethod) args[4]
+        )
+    );
+    private static final ParseField ALTERNATIVE = new ParseField("alternative");
+    private static final ParseField SAMPLING_METHOD = new ParseField("sampling_method");
+    private static final ParseField FRACTIONS = new ParseField("fractions");
+
+    static {
+        PARSER.declareString(ConstructingObjectParser.constructorArg(), BUCKETS_PATH_FIELD);
+        PARSER.declareDoubleArray(ConstructingObjectParser.optionalConstructorArg(), FRACTIONS);
+        PARSER.declareStringArray(ConstructingObjectParser.optionalConstructorArg(), ALTERNATIVE);
+        PARSER.declareField(ConstructingObjectParser.optionalConstructorArg(), p -> {
+            if (p.currentToken() == XContentParser.Token.VALUE_STRING) {
+                return BucketHelpers.GapPolicy.parse(p.text().toLowerCase(Locale.ROOT), p.getTokenLocation());
+            }
+            throw new IllegalArgumentException("Unsupported token [" + p.currentToken() + "]");
+        }, GAP_POLICY, ObjectParser.ValueType.STRING);
+        PARSER.declareField(ConstructingObjectParser.optionalConstructorArg(), p -> {
+            if (p.currentToken() == XContentParser.Token.VALUE_STRING) {
+                return SamplingMethod.fromString(p.text().toLowerCase(Locale.ROOT));
+            }
+            throw new IllegalArgumentException("Unsupported token [" + p.currentToken() + "]");
+        }, SAMPLING_METHOD, ObjectParser.ValueType.STRING);
+    }
+
+    private final double[] fractions;
+    private final EnumSet<Alternative> alternative;
+    private final SamplingMethod samplingMethod;
+    private BucketCountKSTestAggregationBuilder(
+        String name,
+        String bucketsPath,
+        @Nullable List<Double> fractions,
+        @Nullable List<String> alternative,
+        @Nullable BucketHelpers.GapPolicy gapPolicy,
+        @Nullable SamplingMethod samplingMethod
+    ) {
+        super(
+            name,
+            NAME.getPreferredName(),
+            new String[] { bucketsPath },
+            null,
+            gapPolicy == null ? BucketHelpers.GapPolicy.INSERT_ZEROS : gapPolicy
+        );
+        this.fractions = fractions == null ? null : fractions.stream().mapToDouble(Double::doubleValue).toArray();
+        if (alternative == null) {
+            this.alternative = EnumSet.allOf(Alternative.class);
+        } else {
+            if (alternative.isEmpty()) {
```

Comment:
The exact same check happens in `validate()` but it throws a validation error rather than `IllegalArgumentException`.

Should the `gapPolicy` check  below also be in `validate()` (not the null part the bit about policy != INSERT_ZERORS). A quick survey of other agg builder classes does not show a strong convention either way.

Reply:
I am gonna remove the check from `validate()`. All the fields are `final` and I already validate in the ctor.